text/0000-lifecycle-parallel-export.md

```diff
@@ -0,0 +1,213 @@
+
+# Motivation
+[motivation]: #motivation
+
+In some scenario, the app image and the cache image need to be exported at the same time, but this process is serial in the lifecycle, which means that after the app image is exported, we have to wait for the cache image to be exported. But we don’t need to wait for the export of cache image, only after the app is exported, we can continue to next steps (distribution and deployment).
+
+So we can try to parallelize this step ([lifecycle#1167](https://github.com/buildpacks/lifecycle/pull/1167)) and compare it with serial exporting. After testing the build on some applications, this modification can shorten the export time.
+
+- Java (app image is 202.361MB, cache image is 157.525MB, with one same layer: 107.648MB):
+    - Before: total 18.34s, app 8.96s, cache 9.38s
+    - After: total 14.70s, app 11.42s, cache 13.93s
+    - app image layers: 0+1.103MB+15.153MB+107.648MB+49.953MB+0+0+28.502MB
+    - cache image layers: 9.411MB+40.465MB+107.648MB
+
+- Go (app image is 114.273MB, cache is 175.833MB, no same layer):
+    - Before: total 16.57s, app 5.92s, cache 10.65s
+    - After: total 12.02s, app 7.31s, cache 11.48s
+    - app image layers: 0+1MB+25.72MB+8.993MB+49.953MB+0+0+28.502MB
+    - cache image layers: 70.87MB+104.964MB
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is to add a new capability to the lifecycle (enabled by configuration) to export app image and cache image to registry in parallel.
+
+The target personas affected by this change are:
+
+ - **buildpack user**: they will experience higher disk I/O or network pressure if this feature is enabled by default.
+ - **Platform implementors**: they will choose parallel or serial export, to suit how the platform works. Serial export helps to get app image faster, while parallel export can complete the build process faster.
+
+This proposal, in addition to acceleration, has the greatest impact on the export time of app image. This will lead to resource competition, that is, when the app image and cache image are exported at the same time, the app image export time will become longer. For some users, they only care about the export time of the app image, but not the overall optimization effect, and enabling this capability will affect their performance. Therefore, this ability is optional.
+
+The flag name, refer to other boolean flag like `daemon`, `layout`, I think it can be named `parallel`. And then the usage is just like this:
+```
+/cnb/lifecycle/exporter \
+  [-analyzed <analyzed>] \
+  [-app <app>] \
+  [-cache-dir <cache-dir>] \
+  [-cache-image <cache-image>] \
+  [-daemon] \ # sets <daemon>
+  [-extended <extended>] \
+  [-gid <gid>] \
+  [-group <group>] \
+  [-launch-cache <launch-cache> ] \
+  [-launcher <launcher> ] \
+  [-launcher-sbom <launcher-sbom> ] \
+  [-layers <layers>] \
+  [-layout] \ # sets <layout>
+  [-layout-dir] \ # sets <layout-dir>
+  [-log-level <log-level>] \
+  [-parallel] \ # sets <parallel>
```

Comment:
Suggestion from working group - when these flags are combined in the `creator` `-parallel-export` might make more sense. However `-parallel-export` might look funny as a flag to the exporter.

Perhaps `-parallel-export` as a flag to the creator and `-parallel` as a flag to the exporter? I don't have a strong opinion either way, but thought I'd surface this idea.

Reply:
@ESWZY @jabrown85 perhaps you could opine here? Then I'll move this one back into voting. We already have enough approvals.